### PR TITLE
[Merged by Bors] - feat(CategoryTheory): pushforward pullback adjunction for `P.Over Q X`

### PR DESCRIPTION
**Pull request by chrisflav**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/MorphismProperty/Comma.lean [hunk 1]
```diff
@@ -203,6 +211,102 @@ def forgetFullyFaithful : (forget L R P ⊤ ⊤).FullyFaithful where
...
+  obj X :=
+    { left := X.left
+      right := X.right
+      hom := l.app X.left ≫ X.hom
+      prop := hl X }
+  map {X Y} f :=
+    { left := f.left
+      right := f.right
+      prop_hom_left := f.prop_hom_left
+      prop_hom_right := f.prop_hom_right }
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/Comma.lean:266]: The following definitions is defeq to yours and it also generates the expected simp lemmas:
```suggestion
@[simps! obj_left obj_right map_left map_right]
def mapLeft(l : L₁ ⟶ L₂) (hl : ∀ X : P.Comma L₂ R Q W, P (l.app X.left ≫ X.hom)) :
    P.Comma L₂ R Q W ⥤ P.Comma L₁ R Q W :=
  lift (forget _ _ _ _ _ ⋙ CategoryTheory.Comma.mapLeft R l) hl
    (fun f ↦ f.prop_hom_left) (fun f ↦ f.prop_hom_right)
```
The same should be done for `mapRight`.

File: Mathlib/CategoryTheory/MorphismProperty/Comma.lean [hunk 2]
```diff
@@ -203,6 +211,102 @@ def forgetFullyFaithful : (forget L R P ⊤ ⊤).FullyFaithful where
...
+end
+
+section Functoriality
+
+variable {L R P Q W}
+variable {L₁ L₂ L₃ : A ⥤ T} {R₁ R₂ R₃ : B ⥤ T}
+
+/-- Lift a functor `F : C ⥤ Comma L R` to the subcategory `P.Comma L R Q W` under
+suitable assumptions on `F`. -/
+def lift {C : Type*} [Category C] (F : C ⥤ Comma L R)
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/Comma.lean:239]:
```suggestion
@[simps obj_toComma map_hom]
def lift {C : Type*} [Category C] (F : C ⥤ Comma L R)
```

File: Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+section Map
+
+variable {P} [P.IsStableUnderComposition] (hPf : P f)
+
+variable {f}
+
+/-- If `P` is stable under composition and `f : X ⟶ Y` satisfies `P`,
+this is the functor `P.Over Q X ⥤ P.Over Q Y` given by composing with `f`. -/
+def Over.map : P.Over Q X ⥤ P.Over Q Y :=
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean:37]:
```suggestion
@[simps! obj_left obj_hom map_left]
def Over.map : P.Over Q X ⥤ P.Over Q Y :=
```

File: Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  · simp [map, Comma.mapRight, CategoryTheory.Comma.mapRight]
+  · intro U V k
+    ext
+    simp
+
+/-- `Over.map` commutes with composition. -/
+@[simps!]
+def Over.mapComp {X Y Z : T} {f : X ⟶ Y} (hf : P f) {g : Y ⟶ Z} (hg : P g) :
+    map Q (P.comp_mem f g hf hg) ≅ map Q hf ⋙ map Q hg :=
+  eqToIso (Over.mapComp_eq Q hf hg)
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean:59]: Unless you intend to use this when `Q` does not respect isos, I would favour the following (which avoids `eqToIso`):
```suggestion
@[simps! hom_app_left inv_app_left]
def Over.mapComp {X Y Z : T} {f : X ⟶ Y} (hf : P f) {g : Y ⟶ Z} (hg : P g) [Q.RespectsIso] :
    map Q (P.comp_mem f g hf hg) ≅ map Q hf ⋙ map Q hg :=
  NatIso.ofComponents (fun X ↦ Over.isoMk (Iso.refl _))
```

File: Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean (hunk 3 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean:37]: It seems there is a linting issue with the generated `map_left`. Stating the lemmas separately as you did is a workaround, but the best approach would be the point out the problem on Zulip (tagging Floris who may understand better what is happening).

File: Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+@[simp]
+lemma Over.pullback_map_left {A B : P.Over Q Y} (g : A ⟶ B) :
+    ((Over.pullback P Q f).map g).left = pullback.lift
+      (pullback.fst A.hom f ≫ g.left) (pullback.snd A.hom f) (by simp [pullback.condition]) :=
+  rfl
+
+variable {P} {Q}
+
+/-- `Over.pullback` commutes with  composition. -/
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean:87]:
```suggestion
/-- `Over.pullback` commutes with composition. -/
```

File: Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        pullback.fst A.hom (f ≫ g) := by
+  simp
+
+/-- If `f = g`, then base change along `f` is naturally isomorphic to base change along `g`. -/
+noncomputable def Over.pullbackCongr {X Y : T} {f g : X ⟶ Y} (h : f = g) :
+    Over.pullback P Q f ≅ Over.pullback P Q g :=
+  NatIso.ofComponents (fun X ↦ eqToIso (by rw [h]))
+
+@[reassoc (attr := simp)]
+lemma Over.pullbackCongr_left_fst {X Y : T} {f g : X ⟶ Y} (h : f = g) (A : P.Over Q Y) :
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean:107]:
```suggestion
lemma Over.pullbackCongr_hom_app_left_fst {X Y : T} {f g : X ⟶ Y} (h : f = g) (A : P.Over Q Y) :
```

File: Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean (hunk 3 above)
Comment (chrisflav) [reply Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean:38]: After spending two hours on minimizing and debugging this, I found the issue: It is the same as this one: https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/!4.234183.20.60simps.60.20maximum.20recursion.20depth. I removed `@[simp]` from the two bad simp lemmas (that can be proven by `simp`, but don't seem to get flagged by `simpNF`) and now everything works.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Comment by chrisflav** [comment]
Thanks for the review @joelriou!

**Pull request closed without merging**